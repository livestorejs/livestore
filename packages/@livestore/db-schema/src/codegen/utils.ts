import os from 'node:os'
import path from 'node:path'

import dprint from 'dprint-node'

export const pretty = (code: string): string => {
  const tmpPath = os.tmpdir()
  const filePath = path.join(tmpPath, 'tmp.ts')

  return dprint.format(filePath, code, {
    semiColons: 'asi',
    quoteStyle: 'preferSingle',
  })
}
