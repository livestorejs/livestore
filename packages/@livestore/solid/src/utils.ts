import type { Accessor } from 'solid-js'

export type MakeOptional<T, TKeys extends keyof T> = Omit<T, TKeys> & { [TKey in TKeys]: T[TKey] | undefined }

export type AccessorMaybe<T> = Accessor<T> | T

export function resolve<T>(value: AccessorMaybe<T>): T {
  if (typeof value === 'function') {
    return (value as Accessor<T>)()
  }
  return value
}
