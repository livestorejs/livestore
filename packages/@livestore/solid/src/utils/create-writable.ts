import { type Accessor, createMemo, createSignal } from 'solid-js'

/**
 * Creates a signal that can also be updated if the accessor updates.
 * @param accessor - A function that returns the initial value for the signal
 * @returns A tuple of [getter, setter] similar to createSignal, but the signal is recreated when the accessor's value changes
 * @remarks This is useful when you need a signal that can be manually set but also updates when its source value changes
 * @example
 * ```ts
 * const [width, setWidth] = createWritable(() => config().width);
 * // width() will return config().width initially
 * // setWidth(100) will update the signal to 100
 * // If config().width changes, a new signal width() is updated to the new value
 * ```
 */
export function createWritable<T>(accessor: Accessor<T>) {
  // eslint-disable-next-line solid/reactivity
  const signal = createMemo(() => createSignal(accessor()))
  const get = () => signal()[0]()
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const set = (v: any) => signal()[1](v)
  return [get, set] as ReturnType<typeof createSignal<T>>
}
