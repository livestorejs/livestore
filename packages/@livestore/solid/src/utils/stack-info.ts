import { extractStackInfoFromStackTrace, type StackInfo } from '@livestore/livestore'

export const originalStackLimit = Error.stackTraceLimit

export const useStackInfo = (): StackInfo => {
  Error.stackTraceLimit = 10

  const stack = new Error().stack!
  Error.stackTraceLimit = originalStackLimit
  return extractStackInfoFromStackTrace(stack)
}
