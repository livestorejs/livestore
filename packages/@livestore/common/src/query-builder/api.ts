import type { GetValForKey } from '@livestore/utils'
import { type Option, Predicate, type Schema } from '@livestore/utils/effect'

import type { SessionIdSymbol } from '../adapter-types.js'
import type { QueryInfo } from '../query-info.js'
import type { DbSchema } from '../schema/index.js'
import type { SqliteDsl } from '../schema/table-def.js'
import type { SqlValue } from '../util.js'

export type QueryBuilderAst = QueryBuilderAst.SelectQuery | QueryBuilderAst.CountQuery | QueryBuilderAst.RowQuery

export namespace QueryBuilderAst {
  export type SelectQuery = {
    readonly _tag: 'SelectQuery'
    readonly columns: string[]
    readonly pickFirst: false | { fallback: () => any }
    readonly select: {
      columns: ReadonlyArray<string>
    }
    readonly orderBy: ReadonlyArray<OrderBy>
    readonly offset: Option.Option<number>
    readonly limit: Option.Option<number>
    readonly tableDef: DbSchema.TableDefBase
    readonly where: ReadonlyArray<QueryBuilderAst.Where>
    readonly resultSchemaSingle: Schema.Schema<any>
  }

  export type CountQuery = {
    readonly _tag: 'CountQuery'
    readonly tableDef: DbSchema.TableDefBase
    readonly where: ReadonlyArray<QueryBuilderAst.Where>
    readonly resultSchema: Schema.Schema<number, ReadonlyArray<{ count: number }>>
  }

  export type RowQuery = {
    readonly _tag: 'RowQuery'
    readonly tableDef: DbSchema.TableDefBase
    readonly id: string | SessionIdSymbol
    readonly insertValues: Record<string, unknown>
  }

  export type Where = {
    readonly col: string
    readonly op: QueryBuilder.WhereOps
    readonly value: unknown
  }

  export type OrderBy = {
    readonly col: string
    readonly direction: 'asc' | 'desc'
  }
}

export const QueryBuilderAstSymbol = Symbol.for('QueryBuilderAst')
export type QueryBuilderAstSymbol = typeof QueryBuilderAstSymbol
export const TypeId = Symbol.for('QueryBuilder')
export type TypeId = typeof TypeId

export const isQueryBuilder = (value: unknown): value is QueryBuilder<any, any, any> =>
  Predicate.hasProperty(value, TypeId)

export type QueryBuilder<
  TResult,
  TTableDef extends DbSchema.TableDefBase,
  /** Used to gradually remove features from the API based on the query context */
  TWithout extends QueryBuilder.ApiFeature = never,
  TQueryInfo extends QueryInfo = QueryInfo.None,
> = {
  readonly [TypeId]: TypeId
  readonly [QueryBuilderAstSymbol]: QueryBuilderAst
  readonly asSql: () => { query: string; bindValues: SqlValue[] }
  readonly toString: () => string
} & Omit<QueryBuilder.ApiFull<TResult, TTableDef, TWithout, TQueryInfo>, TWithout>

export namespace QueryBuilder {
  export type Any = QueryBuilder<any, any, any, any>
  export type WhereOps = WhereOps.Equality | WhereOps.Order | WhereOps.Like | WhereOps.In

  export namespace WhereOps {
    export type Equality = '=' | '!='
    export type Order = '<' | '>' | '<=' | '>='
    export type Like = 'LIKE' | 'NOT LIKE' | 'ILIKE' | 'NOT ILIKE'
    export type In = 'IN' | 'NOT IN'

    export type SingleValue = Equality | Order | Like
    export type MultiValue = In
  }

  export type ApiFeature = 'select' | 'where' | 'count' | 'orderBy' | 'offset' | 'limit' | 'first' | 'row'

  export type WhereParams<TTableDef extends DbSchema.TableDefBase> = Partial<{
    [K in keyof TTableDef['sqliteDef']['columns']]:
      | TTableDef['sqliteDef']['columns'][K]['schema']['Type']
      | { op: QueryBuilder.WhereOps.SingleValue; value: TTableDef['sqliteDef']['columns'][K]['schema']['Type'] }
      | {
          op: QueryBuilder.WhereOps.MultiValue
          value: ReadonlyArray<TTableDef['sqliteDef']['columns'][K]['schema']['Type']>
        }
      | undefined
  }>

  export type OrderByParams<TTableDef extends DbSchema.TableDefBase> = ReadonlyArray<{
    col: keyof TTableDef['sqliteDef']['columns'] & string
    direction: 'asc' | 'desc'
  }>

  export type ApiFull<
    TResult,
    TTableDef extends DbSchema.TableDefBase,
    TWithout extends ApiFeature,
    TQueryInfo extends QueryInfo,
  > = {
    /**
     * `SELECT *` is the default
     *
     * Example:
     * ```ts
     * db.todos.select('id', 'text', 'completed')
     * db.todos.select('id', { pluck: true })
     * ```
     */
    readonly select: {
      <TColumn extends keyof TTableDef['sqliteDef']['columns'] & string, TPluck extends boolean = false>(
        column: TColumn,
        options?: { pluck: TPluck },
      ): QueryBuilder<
        TPluck extends true
          ? ReadonlyArray<TTableDef['sqliteDef']['columns'][TColumn]['schema']['Type']>
          : ReadonlyArray<{
              readonly [K in TColumn]: TTableDef['sqliteDef']['columns'][K]['schema']['Type']
            }>,
        TTableDef,
        TWithout | 'row' | 'select',
        TQueryInfo
      >
      <TColumns extends keyof TTableDef['sqliteDef']['columns'] & string>(
        ...columns: TColumns[]
        // TODO also support arbitrary SQL selects
        // params: QueryBuilderSelectParams,
      ): QueryBuilder<
        ReadonlyArray<{
          readonly [K in TColumns]: TTableDef['sqliteDef']['columns'][K]['schema']['Type']
        }>,
        TTableDef,
        TWithout | 'row' | 'select' | 'count',
        TQueryInfo
      >
    }

    /**
     * Notes:
     * - All where clauses are `AND`ed together by default.
     * - `null` values only support `=` and `!=` which is translated to `IS NULL` and `IS NOT NULL`.
     *
     * Example:
     * ```ts
     * db.todos.where('completed', true)
     * db.todos.where('completed', '!=', true)
     * db.todos.where({ completed: true })
     * db.todos.where({ completed: { op: '!=', value: true } })
     * ```
     *
     * TODO: Also support `OR`
     */
    readonly where: {
      <TParams extends QueryBuilder.WhereParams<TTableDef>>(
        params: TParams,
      ): QueryBuilder<TResult, TTableDef, TWithout | 'row' | 'select', TQueryInfo>
      <TColName extends keyof TTableDef['sqliteDef']['columns']>(
        col: TColName,
        value: TTableDef['sqliteDef']['columns'][TColName]['schema']['Type'],
      ): QueryBuilder<TResult, TTableDef, TWithout | 'row' | 'select', TQueryInfo>
      <TColName extends keyof TTableDef['sqliteDef']['columns']>(
        col: TColName,
        op: QueryBuilder.WhereOps,
        value: TTableDef['sqliteDef']['columns'][TColName]['schema']['Type'],
      ): QueryBuilder<TResult, TTableDef, TWithout | 'row' | 'select', TQueryInfo>
    }

    /**
     * Example:
     * ```ts
     * db.todos.count()
     * db.todos.count().where('completed', true)
     * ```
     */
    readonly count: () => QueryBuilder<
      number,
      TTableDef,
      TWithout | 'row' | 'count' | 'select' | 'orderBy' | 'first' | 'offset' | 'limit',
      TQueryInfo
    >

    /**
     * Example:
     * ```ts
     * db.todos.orderBy('createdAt', 'desc')
     * ```
     */
    readonly orderBy: {
      <TColName extends keyof TTableDef['sqliteDef']['columns'] & string>(
        col: TColName,
        direction: 'asc' | 'desc',
      ): QueryBuilder<TResult, TTableDef, TWithout, TQueryInfo>
      <TParams extends QueryBuilder.OrderByParams<TTableDef>>(
        params: TParams,
      ): QueryBuilder<TResult, TTableDef, TWithout, TQueryInfo>
    }

    /**
     * Example:
     * ```ts
     * db.todos.offset(10)
     * ```
     */
    readonly offset: (
      offset: number,
    ) => QueryBuilder<TResult, TTableDef, TWithout | 'row' | 'offset' | 'orderBy', TQueryInfo>

    /**
     * Example:
     * ```ts
     * db.todos.limit(10)
     * ```
     */
    readonly limit: (
      limit: number,
    ) => QueryBuilder<TResult, TTableDef, TWithout | 'row' | 'limit' | 'offset' | 'first' | 'orderBy', TQueryInfo>

    /**
     * Example:
     * ```ts
     * db.todos.first()
     * ```
     */
    readonly first: <TFallback extends GetSingle<TResult> = never>(options?: {
      fallback?: () => TFallback
    }) => QueryBuilder<
      TFallback | GetSingle<TResult>,
      TTableDef,
      TWithout | 'row' | 'first' | 'orderBy' | 'select' | 'limit' | 'offset' | 'where',
      TQueryInfo
    >

    /**
     *
     */
    readonly row: TTableDef['options']['isSingleton'] extends true
      ? () => QueryBuilder<RowQuery.Result<TTableDef>, TTableDef, QueryBuilder.ApiFeature, QueryInfo.Row>
      : TTableDef['options']['deriveMutations']['enabled'] extends false
        ? (_: 'Error: Need to enable deriveMutations to use row()') => any
        : TTableDef['options']['requiredInsertColumnNames'] extends never
          ? (
              id: string | SessionIdSymbol,
            ) => QueryBuilder<RowQuery.Result<TTableDef>, TTableDef, QueryBuilder.ApiFeature, QueryInfo.Row>
          : <TOptions extends RowQuery.RequiredColumnsOptions<TTableDef>>(
              id: string | SessionIdSymbol,
              opts: TOptions,
            ) => QueryBuilder<RowQuery.Result<TTableDef>, TTableDef, QueryBuilder.ApiFeature, QueryInfo.Row>
  }
}

export namespace RowQuery {
  export type RequiredColumnsOptions<TTableDef extends DbSchema.TableDefBase> = {
    /**
     * Values to be inserted into the row if it doesn't exist yet
     */
    insertValues: Pick<
      SqliteDsl.FromColumns.RowDecodedAll<TTableDef['sqliteDef']['columns']>,
      SqliteDsl.FromColumns.RequiredInsertColumnNames<Omit<TTableDef['sqliteDef']['columns'], 'id'>>
    >
  }

  export type Result<TTableDef extends DbSchema.TableDefBase> = TTableDef['options']['isSingleColumn'] extends true
    ? GetValForKey<SqliteDsl.FromColumns.RowDecoded<TTableDef['sqliteDef']['columns']>, 'value'>
    : SqliteDsl.FromColumns.RowDecoded<TTableDef['sqliteDef']['columns']>

  export type ResultEncoded<TTableDef extends DbSchema.TableDefBase> =
    TTableDef['options']['isSingleColumn'] extends true
      ? GetValForKey<SqliteDsl.FromColumns.RowEncoded<TTableDef['sqliteDef']['columns']>, 'value'>
      : SqliteDsl.FromColumns.RowEncoded<TTableDef['sqliteDef']['columns']>
}

type GetSingle<T> = T extends ReadonlyArray<infer U> ? U : never

// export type QueryBuilderParamRef = { _tag: 'QueryBuilderParamRef' }
// export type QueryBuilderSelectParams = { [key: string]: QueryBuilderSelectParam }
// export type QueryBuilderSelectParam = boolean | ((ref: QueryBuilderParamRef) => QueryBuilder<any, any>)
