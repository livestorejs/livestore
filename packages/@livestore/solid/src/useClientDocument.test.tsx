/** biome-ignore-all lint/a11y/useValidAriaRole: not needed for testing */
/** biome-ignore-all lint/a11y/noStaticElementInteractions: not needed for testing */
import * as LiveStore from '@livestore/livestore'
import { getAllSimplifiedRootSpans, getSimplifiedRootSpan } from '@livestore/livestore/internal/testing-utils'
import { Effect, ReadonlyRecord, Schema } from '@livestore/utils/effect'
import { Vitest } from '@livestore/utils-dev/node-vitest'
import * as otel from '@opentelemetry/api'
import { BasicTracerProvider, InMemorySpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import * as SolidTesting from '@solidjs/testing-library'
import { beforeEach, expect, it } from 'vitest'

import { events, makeTodoMvcSolid, tables } from './__tests__/fixture.js'
import type * as LiveStoreSolid from './mod.js'
import { __resetUseRcResourceCache } from './useRcResource.js'
import { createSignal, For } from 'solid-js'

// const strictMode = process.env.REACT_STRICT_MODE !== undefined

// NOTE running tests concurrently doesn't work with the default global db graph
Vitest.describe('useClientDocument', () => {
  beforeEach(() => {
    __resetUseRcResourceCache()
  })

  Vitest.scopedLive('should update the data based on component key', () =>
    Effect.gen(function* () {
      const { wrapper, store } = yield* makeTodoMvcSolid({})

      const [userId, setUserId] = createSignal('u1')

      const { result } = SolidTesting.renderHook(
        () => {
          const [state, setState, id] = store.useClientDocument(() => tables.userInfo, userId)
          return { state, setState, id }
        },
        { wrapper },
      )

      expect(result.id()).toBe('u1')
      expect(result.state().username).toBe('')
      expect(store.reactivityGraph.getSnapshot({ includeResults: true })).toMatchSnapshot()
      store.commit(tables.userInfo.set({ username: 'username_u2' }, 'u2'))

      setUserId('u2')

      expect(store.reactivityGraph.getSnapshot({ includeResults: true })).toMatchSnapshot()
      expect(result.id()).toBe('u2')
      expect(result.state().username).toBe('username_u2')
    }),
  )

  // TODO add a test that makes sure Solid doesn't re-render when a setter is used to set the same value

  Vitest.scopedLive('should update the data reactively - via setState', () =>
    Effect.gen(function* () {
      const { wrapper, store } = yield* makeTodoMvcSolid({})

      const { result } = SolidTesting.renderHook(
        () => {
          const [state, setState, id] = store.useClientDocument(
            () => tables.userInfo,
            () => 'u1',
          )
          return { state, setState, id }
        },
        { wrapper },
      )

      expect(result.id()).toBe('u1')
      expect(result.state().username).toBe('')

      result.setState({ username: 'username_u1_hello' })

      expect(result.id()).toBe('u1')
      expect(result.state().username).toBe('username_u1_hello')
    }),
  )

  Vitest.scopedLive('should update the data reactively - via raw store commit', () =>
    Effect.gen(function* () {
      const { wrapper, store } = yield* makeTodoMvcSolid({})

      const { result } = SolidTesting.renderHook(
        () => {
          const [state, setState, id] = store.useClientDocument(
            () => tables.userInfo,
            () => 'u1',
          )
          return { state, setState, id }
        },
        { wrapper },
      )

      expect(result.id()).toBe('u1')
      expect(result.state().username).toBe('')

      store.commit(events.UserInfoSet({ username: 'username_u1_hello' }, 'u1'))

      expect(result.id()).toBe('u1')
      expect(result.state().username).toBe('username_u1_hello')
    }),
  )

  Vitest.scopedLive('should work for a larger app', () =>
    Effect.gen(function* () {
      const { wrapper, store } = yield* makeTodoMvcSolid({})

      const allTodos$ = LiveStore.queryDb(
        { query: `select * from todos`, schema: Schema.Array(tables.todos.rowSchema) },
        { label: 'allTodos' },
      )

      let globalSetState: LiveStoreSolid.StateSetters<typeof tables.AppRouterSchema> | undefined
      const AppRouter = () => {
        const [state, setState] = store.useClientDocument(
          () => tables.AppRouterSchema,
          () => 'singleton',
        )

        globalSetState = setState

        return (
          <div>
            <TasksList setTaskId={(taskId) => setState({ currentTaskId: taskId })} />
            <div role="current-id">Current Task Id: {state().currentTaskId ?? '-'}</div>
            {state().currentTaskId ? (
              <TaskDetails id={state().currentTaskId} />
            ) : (
              <div>Click on a task to see details</div>
            )}
          </div>
        )
      }

      const TasksList = (props: { setTaskId: (_: string) => void }) => {
        const allTodos = store.useQuery(() => allTodos$)

        return (
          <div>
            <For each={allTodos()}>{(todo) => <div onClick={() => props.setTaskId(todo.id)}>{todo.id}</div>}</For>
          </div>
        )
      }

      const TaskDetails = (props: { id: string }) => {
        const todo = store.useQuery(() =>
          LiveStore.queryDb(tables.todos.where({ id: props.id }).first(), { deps: props.id }),
        )

        return <div role="content">{JSON.stringify(todo())}</div>
      }

      const { getByRole } = SolidTesting.render(() => <AppRouter />, { wrapper })

      store.commit(events.todoCreated({ id: 't1', text: 'buy milk', completed: false }))

      expect(getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: -"')

      globalSetState!({ currentTaskId: 't1' })

      expect(getByRole('content').innerHTML).toMatchInlineSnapshot(`"{"id":"t1","text":"buy milk","completed":false}"`)

      expect(getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t1"')

      store.commit(
        events.todoCreated({ id: 't2', text: 'buy eggs', completed: false }),
        events.AppRouterSet({ currentTaskId: 't2' }),
        events.todoCreated({ id: 't3', text: 'buy bread', completed: false }),
      )

      expect(getByRole('current-id').innerHTML).toMatchInlineSnapshot('"Current Task Id: t2"')
    }),
  )

  Vitest.scopedLive('should work for a useClientDocument query chained with a useTemporary query', () =>
    Effect.gen(function* () {
      const { store, wrapper } = yield* makeTodoMvcSolid({})

      store.commit(
        events.todoCreated({ id: 't1', text: 'buy milk', completed: false }),
        events.todoCreated({ id: 't2', text: 'buy bread', completed: false }),
      )

      const [userId, setUserId] = createSignal('u1')

      const { result } = SolidTesting.renderHook(
        () => {
          const [_row, _setRow, _id, rowState$] = store.useClientDocument(() => tables.userInfo, userId)
          const todos = store.useQuery(
            () =>
              LiveStore.queryDb(
                (get) => tables.todos.where('text', 'LIKE', `%${get(rowState$()).text}%`),
                // TODO find a way where explicit `userId` is not needed here
                // possibly by automatically understanding the `get(rowState$)` dependency
                { label: 'todosFiltered', deps: userId() },
              ),
            // TODO introduce a `deps` array which is only needed when a query is parametric
          )

          return { todos }
        },
        { wrapper },
      )

      expect(result.todos().length).toBe(2)

      // Set text filter for u2 and test with second user
      store.commit(events.UserInfoSet({ username: 'username_u2', text: 'milk' }, 'u2'))

      setUserId('u2')

      expect(result.todos().length).toBe(1)
    }),
  )

  Vitest.scopedLive('kv client document overwrites value (Schema.Any, no partial merge)', () =>
    Effect.gen(function* () {
      const { wrapper, store } = yield* makeTodoMvcSolid({})

      const { result } = SolidTesting.renderHook(
        () => {
          const [state, setState] = store.useClientDocument(
            () => tables.kv,
            () => 'k1',
          )
          return { state, setState, id: () => 'k1' }
        },
        { wrapper },
      )

      expect(result.id()).toBe('k1')
      expect(result.state()).toBe(null)

      result.setState(1)
      expect(result.state()).toEqual(1)

      result.setState({ b: 2 })
      expect(result.state()).toEqual({ b: 2 })
    }),
  )

  Vitest.describe('otel', () => {
    it('should update the data based on component key', async () => {
      const exporter = new InMemorySpanExporter()

      const provider = new BasicTracerProvider({
        spanProcessors: [new SimpleSpanProcessor(exporter)],
      })

      const otelTracer = provider.getTracer(`testing-solid`)

      const span = otelTracer.startSpan('test-root')
      const otelContext = otel.trace.setSpan(otel.context.active(), span)

      await Effect.gen(function* () {
        const { wrapper, store } = yield* makeTodoMvcSolid({
          otelContext,
          otelTracer,
        })

        // Test with first user
        const { result: result1 } = SolidTesting.renderHook(
          () => {
            const [state, setState, id] = store.useClientDocument(
              () => tables.userInfo,
              () => 'u1',
            )
            return { state, setState, id }
          },
          { wrapper },
        )

        expect(result1.id()).toBe('u1')
        expect(result1.state().username).toBe('')

        // For u2 we'll make sure that the row already exists,
        // so the lazy `insert` will be skipped
        store.commit(events.UserInfoSet({ username: 'username_u2' }, 'u2'))

        // Test with second user (new hook instance)
        const { result: result2 } = SolidTesting.renderHook(
          () => {
            const [state, setState, id] = store.useClientDocument(
              () => tables.userInfo,
              () => 'u2',
            )
            return { state, setState, id }
          },
          { wrapper },
        )

        expect(result2.id()).toBe('u2')
        expect(result2.state().username).toBe('username_u2')

        // Note: Solid testing library cleanup happens automatically
        span.end()
      }).pipe(Effect.scoped, Effect.tapCauseLogPretty, Effect.runPromise)

      await provider.forceFlush()

      const mapAttributes = (attributes: otel.Attributes) => {
        return ReadonlyRecord.map(attributes, (val, key) => {
          if (key === 'code.stacktrace') {
            return '<STACKTRACE>'
          } else if (key === 'firstStackInfo') {
            const stackInfo = JSON.parse(val as string) as LiveStore.StackInfo
            // stackInfo.frames.shift() // Removes `renderHook.wrapper` from the stack
            stackInfo.frames.forEach((_) => {
              if (_.name.includes('renderHook.wrapper')) {
                _.name = 'renderHook.wrapper'
              }
              _.filePath = '__REPLACED_FOR_SNAPSHOT__'
            })
            return JSON.stringify(stackInfo)
          }
          return val
        })
      }

      expect(getSimplifiedRootSpan(exporter, 'createStore', mapAttributes)).toMatchSnapshot()
      expect(getAllSimplifiedRootSpans(exporter, 'LiveStore:commit', mapAttributes)).toMatchSnapshot()

      await provider.shutdown()
    })
  })
})
