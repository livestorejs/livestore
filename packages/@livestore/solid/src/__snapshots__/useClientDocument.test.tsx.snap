// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`useClientDocument > otel > should update the data based on component key 1`] = `
{
  "_name": "createStore",
  "attributes": {
    "debugInstanceId": "test",
    "storeId": "default",
  },
  "children": [
    {
      "_name": "livestore.in-memory-db:execute",
      "attributes": {
        "sql.query": "
      PRAGMA page_size=32768;
      PRAGMA cache_size=10000;
      PRAGMA synchronous='OFF';
      PRAGMA temp_store='MEMORY';
      PRAGMA foreign_keys='ON'; -- we want foreign key constraints to be enforced
    ",
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "@livestore/common:LeaderSyncProcessor:push",
      "attributes": {
        "batch": "undefined",
        "batchSize": 1,
      },
    },
    {
      "_name": "client-session-sync-processor:pull",
      "attributes": {
        "code.stacktrace": "<STACKTRACE>",
        "span.label": "⚠︎ Interrupted",
        "status.interrupted": true,
      },
    },
    {
      "_name": "LiveStore:sync",
    },
    {
      "_name": "LiveStore:commits",
    },
    {
      "_name": "LiveStore:queries",
      "children": [
        {
          "_name": "LiveStore:useClientDocument:() => (0,__vite_ssr_import_8__.resolve)(table).sqliteDef.name:() => id() === __vite_ssr_import_0__.SessionIdSymbol ? store.clientSession.sessionId : id()",
          "attributes": {
            "firstStackInfo": "{"frames":[{"name":"Store.useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"},{"name":"useClientDocument","filePath":"__REPLACED_FOR_SNAPSHOT__"}]}",
            "label": "SELECT * FROM 'UserInfo' WHERE id = ?",
          },
          "children": [
            {
              "_name": "sql-in-memory-select",
              "attributes": {
                "sql.cached": false,
                "sql.query": "SELECT 1 FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 0,
              },
            },
            {
              "_name": "db:SELECT * FROM 'UserInfo' WHERE id = ?",
              "attributes": {
                "livestore.debugRefreshReason": "useQuery:initial-run:UserInfo.get:u1",
                "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                "sql.rowsCount": 1,
              },
              "children": [
                {
                  "_name": "sql-in-memory-select",
                  "attributes": {
                    "sql.cached": false,
                    "sql.query": "SELECT * FROM 'UserInfo' WHERE id = ?",
                    "sql.rowsCount": 1,
                  },
                },
              ],
            },
          ],
        },
      ],
    },
  ],
}
`;

exports[`useClientDocument > otel > should update the data based on component key 2`] = `
[
  {
    "_name": "LiveStore:commit",
    "attributes": {
      "livestore.commitLabel": "UserInfo.set:u1",
      "livestore.eventTags": "[
  "UserInfoSet"
]",
      "livestore.eventsCount": 1,
    },
    "children": [
      {
        "_name": "client-session-sync-processor:push",
        "attributes": {
          "batchSize": 1,
          "eventCounts": "{
  "UserInfoSet": 1
}",
          "mergeResultTag": "advance",
        },
        "children": [
          {
            "_name": "client-session-sync-processor:materialize-event",
            "children": [
              {
                "_name": "livestore.in-memory-db:execute",
                "attributes": {
                  "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(json_set(value, ?, json(?)), ?, json(?))
    ",
                },
              },
            ],
          },
        ],
      },
    ],
  },
  {
    "_name": "LiveStore:commit",
    "attributes": {
      "livestore.eventTags": "[
  "UserInfoSet"
]",
      "livestore.eventsCount": 1,
    },
    "children": [
      {
        "_name": "client-session-sync-processor:push",
        "attributes": {
          "batchSize": 1,
          "eventCounts": "{
  "UserInfoSet": 1
}",
          "mergeResultTag": "advance",
        },
        "children": [
          {
            "_name": "client-session-sync-processor:materialize-event",
            "children": [
              {
                "_name": "livestore.in-memory-db:execute",
                "attributes": {
                  "sql.query": "
      INSERT INTO 'UserInfo' (id, value)
      VALUES (?, ?)
      ON CONFLICT (id) DO UPDATE SET value = json_set(value, ?, json(?))
    ",
                },
              },
            ],
          },
        ],
      },
    ],
  },
]
`;

exports[`useClientDocument > should update the data based on component key 1`] = `
{
  "atoms": [
    {
      "_tag": "ref",
      "id": "node-1",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:todos",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-2",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:app",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-3",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:UserInfo",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 1,
      "sub": [],
      "super": [
        "node-6",
      ],
    },
    {
      "_tag": "ref",
      "id": "node-4",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:AppRouter",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-5",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:Kv",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "thunk",
      "id": "node-6",
      "isDestroyed": false,
      "isDirty": false,
      "label": "UserInfo.get:u1:results",
      "meta": {
        "liveStoreThunkType": "db.result",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "{"username":"","text":""}",
      },
      "recomputations": 1,
      "sub": [
        "node-3",
      ],
      "super": [
        "node-7",
      ],
    },
  ],
  "deferredEffects": [],
  "effects": [
    {
      "_tag": "effect",
      "id": "node-7",
      "invocations": 1,
      "isDestroyed": false,
      "label": "subscribe:UserInfo.get:u1",
      "sub": [
        "node-6",
      ],
    },
  ],
}
`;

exports[`useClientDocument > should update the data based on component key 2`] = `
{
  "atoms": [
    {
      "_tag": "ref",
      "id": "node-1",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:todos",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-2",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:app",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-3",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:UserInfo",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 2,
      "sub": [],
      "super": [
        "node-8",
      ],
    },
    {
      "_tag": "ref",
      "id": "node-4",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:AppRouter",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "ref",
      "id": "node-5",
      "isDestroyed": false,
      "isDirty": false,
      "label": "tableRef:Kv",
      "meta": {
        "liveStoreRefType": "table",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "null",
      },
      "refreshes": 0,
      "sub": [],
      "super": [],
    },
    {
      "_tag": "thunk",
      "id": "node-8",
      "isDestroyed": false,
      "isDirty": false,
      "label": "UserInfo.get:u2:results",
      "meta": {
        "liveStoreThunkType": "db.result",
      },
      "previousResult": {
        "_tag": "Some",
        "value": "{"username":"username_u2","text":""}",
      },
      "recomputations": 2,
      "sub": [
        "node-3",
      ],
      "super": [
        "node-9",
      ],
    },
  ],
  "deferredEffects": [],
  "effects": [
    {
      "_tag": "effect",
      "id": "node-9",
      "invocations": 1,
      "isDestroyed": false,
      "label": "subscribe:UserInfo.get:u2",
      "sub": [
        "node-8",
      ],
    },
  ],
}
`;
